### DISCUSSION TITLE
How to draw fine lines correctly?

### DISCUSSION THREAD

**Discussion by dbuzinin** [pixijs/pixijs#8890] [Q&A]
Is there any way to draw fine lines (1px) without losing details?

I made a minimal example of my application, in fact, there can be drawn about 1000 shapes and even more connections between them.

When using cacheasbitmap or Sprite and scaling down the parent container, the fine lines start to lose detail. For example, a rectangle has only 2 sides out of 4 visible.

https://codesandbox.io/s/pixi-playground-forked-n5wo21?file=/src/index.js

![paper](https://user-images.githubusercontent.com/75254600/203570176-7ac3e917-07b8-40d8-9bd8-7a43854ef8f6.png)
![codesandbox](https://user-images.githubusercontent.com/75254600/203570591-5a70e034-2d72-403f-baff-aa4dca6da8ee.png)

### Steps to Reproduce
Rotate the mouse wheel over the rectangle
### Environment

- **`pixi.js` version**:  6.5.3_
- **Browser & Version**:  Chrome Latest
- **OS & Version**: Windows 10
- **Running Example**:  https://codesandbox.io/s/pixi-playground-forked-n5wo21?file=/src/index.js

**Comment by bigtimebuddy** [comment]
Your main problem here is cacheAsBitmap. That converts your graphics to a Texture at the size the Graphics is drawn. Because Textures use linear interpolation when they are scaled, at extreme scales, things start to look a little chunky (in this case, losing a line). Consider not using cacheAsBitmap or try to disable it while you are interacting, and only enabling it after viewport finishes scaling.

Also, we have a plugin called [Smooth Graphics](https://github.com/pixijs/graphics-smooth) was developed for rendering sub-pixel widths and doing better anti-aliasing for strokes. It's pretty good and can perform better at small sizes.
https://codesandbox.io/s/pixi-playground-forked-yhxqb4

Another alternative you may want to consider is the `native: true` option for lineStyle. These lines are rendered on the GPU and do not consider world scale at all. This means that they basically ignore scale and always render at 1px (can only render 1px widths, however).
https://codesandbox.io/s/pixi-playground-forked-ioi6p1?file=/src/index.js

**Reply by dbuzinin** [reply]
Thanks for the quick response! I'll try the options you suggested.

I showed an example with cacheAsBitmap because it gives the same result as when using a sprite from the generated texture.

Because the page elements, as in the screenshot, consist of several shapes, so first I draw them using Graphics and then generate a texture. A 1px-thick stroke can be for 1000 elements on a page (this is how the element selection is displayed). So I also add it to the texture.

If I use Container for an element and there is an element texture and Graphics (stroke) inside it, then there are no problems with the line thickness, but there are performance problems (when dragging the parent container "viewport")

**Reply by MellowCo** [reply]
> Your main problem here is cacheAsBitmap. That converts your graphics to a Texture at the size the Graphics is drawn. Because Textures use linear interpolation when they are scaled, at extreme scales, things start to look a little chunky (in this case, losing a line). Consider not using cacheAsBitmap or try to disable it while you are interacting, and only enabling it after viewport finishes scaling.
> 
> Also, we have a plugin called [Smooth Graphics](https://github.com/pixijs/graphics-smooth) was developed for rendering sub-pixel widths and doing better anti-aliasing for strokes. It's pretty good and can perform better at small sizes. [codesandbox.io/s/pixi-playground-forked-yhxqb4](https://codesandbox.io/s/pixi-playground-forked-yhxqb4)
> 
> Another alternative you may want to consider is the `native: true` option for lineStyle. These lines are rendered on the GPU and do not consider world scale at all. This means that they basically ignore scale and always render at 1px (can only render 1px widths, however). [codesandbox.io/s/pixi-playground-forked-ioi6p1?file=/src/index.js](https://codesandbox.io/s/pixi-playground-forked-ioi6p1?file=/src/index.js)

How to adjust the V8 version as it uses `stroke` without `native`
